File: extensions/twitch/src/main/java/app/revanced/extension/twitch/patches/ForceQualityPatch.java
package app.revanced.extension.twitch.patches;

import com.amazonaws.ivs.player.MediaPlayer;
import com.amazonaws.ivs.player.Quality;

import java.util.Set;

import app.revanced.extension.shared.Logger;
import app.revanced.extension.twitch.settings.Settings;

@SuppressWarnings("unused")
public class ForceQualityPatch {
    public static boolean forceMaxQuality(MediaPlayer player, String requested) {
        if (!Settings.FORCE_MAX_QUALITY.get() || player == null) {
            return false;
        }

        if ("audio_only".equals(requested)) {
            return false;
        }

        Set<Quality> qualities = player.getQualities();
        if (qualities == null || qualities.isEmpty()) {
            return false;
        }

        Quality best = null;
        for (Quality quality : qualities) {
            if (best == null
                    || quality.getHeight() > best.getHeight()
                    || (quality.getHeight() == best.getHeight() && quality.getBitrate() > best.getBitrate())) {
                best = quality;
            }
        }

        if (best == null) {
            return false;
        }

        final Quality selected = best;
        player.setAutoQualityMode(false);
        player.setQuality(selected);
        Logger.printDebug(() -> "Forced max quality: " + selected.getName());
        return true;
    }
}

File: patches/src/main/kotlin/app/revanced/patches/twitch/player/quality/ForceQualityPatch.kt
package app.revanced.patches.twitch.player.quality

import app.revanced.patcher.extensions.ExternalLabel
import app.revanced.patcher.extensions.addInstructionsWithLabels
import app.revanced.patcher.extensions.getInstruction
import app.revanced.patcher.patch.bytecodePatch
import app.revanced.patches.all.misc.resources.addResources
import app.revanced.patches.all.misc.resources.addResourcesPatch
import app.revanced.patches.shared.misc.settings.preference.SwitchPreference
import app.revanced.patches.twitch.misc.extension.sharedExtensionPatch
import app.revanced.patches.twitch.misc.settings.PreferenceScreen
import app.revanced.patches.twitch.misc.settings.settingsPatch

@Suppress("unused")
val forceQualityPatch = bytecodePatch(
    name = "Force max video quality",
    description = "Automatically selects the highest available stream quality on playback.",
) {
    dependsOn(
        sharedExtensionPatch,
        settingsPatch,
        addResourcesPatch,
    )

    compatibleWith("tv.twitch.android.app")

    apply {
        addResources("twitch", "player.quality.forceQualityPatch")

        PreferenceScreen.MISC.GENERAL.addPreferences(
            SwitchPreference("revanced_force_max_quality"),
        )

        setQualityMethod.apply {
            addInstructionsWithLabels(
                0,
                """
                    invoke-static { p0, p1 }, Lapp/revanced/extension/twitch/patches/ForceQualityPatch;->forceMaxQuality(Lcom/amazonaws/ivs/player/MediaPlayer;Ljava/lang/String;)Z
                    move-result v0
                    if-eqz v0, :original
                    return-void
                """,
                ExternalLabel("original", getInstruction(0)),
            )
        }
    }
}
